File: extract_tools/extract_tool_windows.go
package main

import (
	"archive/zip"
	"crypto/md5"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/dialog"
	"fyne.io/fyne/v2/widget"
)

// DataExtractor 数据解压器
type DataExtractor struct {
	prefix    string
	outputDir string
	chunks    []string
}

// ExtractorApp GUI 应用
type ExtractorApp struct {
	app              fyne.App
	window           fyne.Window
	prefixSelect     *widget.Select
	outputEntry      *widget.Entry
	verifyCheck      *widget.Check
	logText          *widget.RichText
	logCard          *widget.Card
	logContainer     *fyne.Container
	progressBar      *widget.ProgressBar
	progressLabel    *widget.Label
	startButton      *widget.Button
	toggleLogButton  *widget.Button
	isExtracting     bool
	logVisible       bool
	extractorMutex   sync.Mutex
}

// FindChunks 查找所有 ZIP 分块文件
func (e *DataExtractor) FindChunks() ([]string, error) {
	currentDir, err := os.Getwd()
	if err != nil {
		return nil, err
	}

	matches, err := filepath.Glob(filepath.Join(currentDir, e.prefix+"_*.zip"))
	if err != nil {
		return nil, err
	}

	sort.Strings(matches)
	e.chunks = matches
	return matches, nil
}

// CalculateMD5 计算文件 MD5
func CalculateMD5(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	hash := md5.New()
	if _, err := io.Copy(hash, file); err != nil {
		return "", err
	}

	return fmt.Sprintf("%x", hash.Sum(nil)), nil
}

// ExtractChunk 解压单个 ZIP 文件
func (e *DataExtractor) ExtractChunk(chunkFile string, progressCallback func(int, int)) error {
	reader, err := zip.OpenReader(chunkFile)
	if err != nil {
		return err
	}
	defer reader.Close()

	totalFiles := len(reader.File)

	for idx, file := range reader.File {
		if progressCallback != nil {
			progressCallback(idx+1, totalFiles)
		}

		if err := extractFile(file, e.outputDir); err != nil {
			fmt.Printf("[WARN] 解压文件 %s 时出错: %v\n", file.Name, err)
		}
	}

	return nil
}

// extractFile 解压单个文件
func extractFile(file *zip.File, outputDir string) error {
	filePath := filepath.Join(outputDir, file.Name)

	// 创建目录
	if file.FileInfo().IsDir() {
		os.MkdirAll(filePath, file.Mode())
		return nil
	}

	// 创建父目录
	if err := os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
		return err
	}

	// 解压文件
	outFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, file.Mode())
	if err != nil {
		return err
	}
	defer outFile.Close()

	inFile, err := file.Open()
	if err != nil {
		return err
	}
	defer inFile.Close()

	_, err = io.Copy(outFile, inFile)
	return err
}

// VerifyMD5 验证 MD5 完整性
func (e *DataExtractor) VerifyMD5(prefix string) (bool, int, int) {
	md5File := prefix + ".md5"
	content, err := os.ReadFile(md5File)
	if err != nil {
		fmt.Printf("[!] MD5 文件不存在: %s\n", md5File)
		return true, 0, 0
	}

	verifiedCount := 0
	failedCount := 0

	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		parts := strings.Fields(line)
		if len(parts) < 2 {
			continue
		}

		expectedHash := parts[0]
		filename := strings.Join(parts[1:], " ")

		filePath := filename
		if !filepath.IsAbs(filePath) {
			filePath = filepath.Join(e.outputDir, filename)
		}

		if _, err := os.Stat(filePath); err != nil {
			fmt.Printf("[!] 文件未找到: %s\n", filename)
			failedCount++
			continue
		}

		actualHash, err := CalculateMD5(filePath)
		if err != nil {
			fmt.Printf("[✗] 计算 MD5 失败: %s\n", filename)
			failedCount++
			continue
		}

		if strings.EqualFold(expectedHash, actualHash) {
			fmt.Printf("[✓] %s\n", filename)
			verifiedCount++
		} else {
			fmt.Printf("[✗] %s (MD5 不匹配!)\n", filename)
			fmt.Printf("    期望: %s\n", expectedHash)
			fmt.Printf("    实际: %s\n", actualHash)
			failedCount++
		}
	}

	return failedCount == 0, verifiedCount, failedCount
}

// AutoDetectPrefixes 自动检测数据包前缀
func AutoDetectPrefixes() ([]string, error) {
	currentDir, err := os.Getwd()
	if err != nil {
		return nil, err
	}

	// 匹配所有符合模式的 ZIP 文件
	matches, err := filepath.Glob(filepath.Join(currentDir, "*_*.zip"))
	if err != nil {
		return nil, err
	}

	prefixes := make(map[string]bool)
	for _, match := range matches {
		base := filepath.Base(match)
		// 从文件名中提取前缀：去掉 "_数字.zip" 部分
		parts := strings.Split(base, "_")
		if len(parts) >= 2 {
			prefix := strings.Join(parts[:len(parts)-1], "_")
			prefixes[prefix] = true
		}
	}

	var result []string
	for p := range prefixes {
		result = append(result, p)
	}
	sort.Strings(result)

	return result, nil
}

// NewExtractorApp 创建新的应用程序
func NewExtractorApp() *ExtractorApp {
	return &ExtractorApp{
		app:            app.New(),
		isExtracting:   false,
		extractorMutex: sync.Mutex{},
	}
}

// SetupUI 设置 UI 界面
func (ea *ExtractorApp) SetupUI() {
	ea.window = ea.app.NewWindow("数据包解压工具 v2.0")
	ea.window.Resize(fyne.NewSize(850, 680))

	// ===== 参数配置区域 =====
	ea.prefixSelect = widget.NewSelect([]string{}, func(s string) {})
	ea.prefixSelect.PlaceHolder = "自动检测或手动选择..."

	// 自动检测前缀
	prefixes, _ := AutoDetectPrefixes()
	ea.prefixSelect.Options = prefixes
	if len(prefixes) > 0 {
		ea.prefixSelect.SetSelected(prefixes[0])
	}

	// 输出目录选择
	currentDir, _ := os.Getwd()
	ea.outputEntry = widget.NewEntry()
	ea.outputEntry.SetText(currentDir)
	ea.outputEntry.PlaceHolder = "选择解压目录..."

	// MD5 验证选项
	ea.verifyCheck = widget.NewCheck("验证 MD5 完整性", func(b bool) {})
	ea.verifyCheck.Checked = false

	// ===== 日志显示区域（默认隐藏）=====
	ea.logText = widget.NewRichTextFromMarkdown("")
	ea.logText.Wrapping = fyne.TextWrapBreak

	logScroll := container.NewScroll(ea.logText)
	logScroll.SetMinSize(fyne.NewSize(800, 250))

	ea.logContainer = container.NewVBox(logScroll)
	ea.logCard = widget.NewCard("", "", ea.logContainer)

	// ===== 开始解压按钮（更明显）=====
	ea.startButton = widget.NewButton("▶ 开始解压", func() {
		ea.startExtraction()
	})

	// ===== 进度条区域 =====
	ea.progressBar = widget.NewProgressBar()
	ea.progressBar.Min = 0
	ea.progressBar.Max = 1
	ea.progressLabel = widget.NewLabel("就绪")

	// ===== 其他操作按钮 =====
	ea.toggleLogButton = widget.NewButton("显示日志", func() {
		ea.toggleLog()
	})

	// ===== 构建主界面 =====
	ea.updateMainContent()
}

// updateMainContent 更新主界面内容
func (ea *ExtractorApp) updateMainContent() {
	titleLabel := widget.NewRichTextFromMarkdown("## 📦 数据包解压工具")
	subtitleLabel := widget.NewLabel("Benagen 一键解压专用工具")
	titleContainer := container.NewCenter(
		container.NewVBox(
			titleLabel,
			subtitleLabel,
		),
	)

	prefixLabel := widget.NewLabel("数据包前缀")
	outputLabel := widget.NewLabel("输出目录")
	browseButton := widget.NewButton("浏览", func() {
		ea.browseOutputDir()
	})
	outputContainer := container.NewBorder(nil, nil, nil, browseButton, ea.outputEntry)

	configBox := container.NewVBox(
		container.NewPadded(
			container.NewVBox(
				prefixLabel,
				ea.prefixSelect,
			),
		),
		container.NewPadded(
			container.NewVBox(
				outputLabel,
				outputContainer,
			),
		),
		container.NewPadded(ea.verifyCheck),
	)

	startButtonBox := container.NewPadded(
		container.NewCenter(ea.startButton),
	)

	progressBox := container.NewPadded(
		container.NewVBox(
			ea.progressLabel,
			ea.progressBar,
		),
	)

	otherButtonBox := container.NewPadded(
		container.NewHBox(
			ea.toggleLogButton,
			widget.NewButton("清空日志", func() { ea.clearLog() }),
		),
	)

	var mainContent *fyne.Container
	if ea.logVisible {
		mainContent = container.NewVBox(
			container.NewPadded(titleContainer),
			widget.NewSeparator(),
			configBox,
			widget.NewSeparator(),
			startButtonBox,
			progressBox,
			otherButtonBox,
			widget.NewSeparator(),
			ea.logCard,
		)
		ea.window.SetContent(container.NewScroll(mainContent))
		ea.window.Resize(fyne.NewSize(850, 900))
	} else {
		mainContent = container.NewVBox(
			container.NewPadded(titleContainer),
			widget.NewSeparator(),
			configBox,
			widget.NewSeparator(),
			startButtonBox,
			progressBox,
			otherButtonBox,
		)
		ea.window.SetContent(mainContent)
		ea.window.Resize(fyne.NewSize(850, 680))
	}
}

// toggleLog 切换日志显示
func (ea *ExtractorApp) toggleLog() {
	ea.logVisible = !ea.logVisible
	if ea.logVisible {
		ea.toggleLogButton.SetText("隐藏日志")
	} else {
		ea.toggleLogButton.SetText("显示日志")
	}
	ea.updateMainContent()
}

// addLog 添加日志消息
func (ea *ExtractorApp) addLog(message string) {
	timestamp := time.Now().Format("15:04:05")
	logEntry := fmt.Sprintf("`[%s]` %s  \n", timestamp, message)
	currentText := ea.logText.String()
	ea.logText.ParseMarkdown(currentText + logEntry)
}

// clearLog 清空日志
func (ea *ExtractorApp) clearLog() {
	ea.logText.ParseMarkdown("")
}

// updateProgress 更新进度条
func (ea *ExtractorApp) updateProgress(current, total int, stage string) {
	if total > 0 {
		percentage := float64(current) / float64(total)
		ea.progressBar.SetValue(percentage)
		ea.progressLabel.SetText(fmt.Sprintf("%s: %d/%d", stage, current, total))
	}
}

// browseOutputDir 浏览输出目录
func (ea *ExtractorApp) browseOutputDir() {
	dialog.ShowFolderOpen(func(list fyne.ListableURI, err error) {
		if err == nil && list != nil {
			ea.outputEntry.SetText(list.Path())
		}
	}, ea.window)
}

// startExtraction 开始解压
func (ea *ExtractorApp) startExtraction() {
	ea.extractorMutex.Lock()
	if ea.isExtracting {
		ea.extractorMutex.Unlock()
		dialog.ShowError(fmt.Errorf("解压正在进行中，请稍候"), ea.window)
		return
	}
	ea.isExtracting = true
	ea.extractorMutex.Unlock()

	prefix := ea.prefixSelect.Selected
	outputDir := ea.outputEntry.Text
	verifyMD5 := ea.verifyCheck.Checked

	if prefix == "" {
		dialog.ShowError(fmt.Errorf("请选择数据包前缀"), ea.window)
		ea.extractorMutex.Lock()
		ea.isExtracting = false
		ea.extractorMutex.Unlock()
		return
	}

	if outputDir == "" {
		dialog.ShowError(fmt.Errorf("请选择输出目录"), ea.window)
		ea.extractorMutex.Lock()
		ea.isExtracting = false
		ea.extractorMutex.Unlock()
		return
	}

	ea.startButton.Disable()

	// 在后台线程执行解压
	go ea.extractionWorker(prefix, outputDir, verifyMD5)
}

// extractionWorker 解压工作线程
func (ea *ExtractorApp) extractionWorker(prefix, outputDir string, verifyMD5 bool) {
	defer func() {
		ea.extractorMutex.Lock()
		ea.isExtracting = false
		ea.extractorMutex.Unlock()
		ea.startButton.Enable()
	}()

	// 创建输出目录
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		ea.addLog(fmt.Sprintf("[✗] 创建输出目录失败: %v", err))
		return
	}

	ea.addLog("[INFO] 开始解压...")
	ea.addLog(fmt.Sprintf("[INFO] 数据包前缀: %s", prefix))
	ea.addLog(fmt.Sprintf("[INFO] 输出目录: %s", outputDir))
	ea.addLog("")

	// 创建解压器
	extractor := &DataExtractor{
		prefix:    prefix,
		outputDir: outputDir,
	}

	// 查找文件
	chunks, err := extractor.FindChunks()
	if err != nil {
		ea.addLog(fmt.Sprintf("[✗] 查找文件失败: %v", err))
		return
	}

	if len(chunks) == 0 {
		ea.addLog("[✗] 没有找到 ZIP 文件")
		dialog.ShowError(fmt.Errorf("没有找到数据包文件"), ea.window)
		return
	}

	ea.addLog(fmt.Sprintf("[INFO] 找到 %d 个分块文件:", len(chunks)))
	for _, chunk := range chunks {
		ea.addLog(fmt.Sprintf("  - %s", filepath.Base(chunk)))
	}
	ea.addLog("")

	// 解压所有文件
	success := true
	for idx, chunk := range chunks {
		ea.addLog(fmt.Sprintf("[INFO] 正在解压第 %d/%d 个文件...", idx+1, len(chunks)))

		if err := extractor.ExtractChunk(chunk, func(current, total int) {
			ea.updateProgress(current, total, fmt.Sprintf("解压文件 (%d/%d)", idx+1, len(chunks)))
		}); err != nil {
			ea.addLog(fmt.Sprintf("[✗] 解压失败 %s: %v", filepath.Base(chunk), err))
			success = false
		} else {
			ea.addLog(fmt.Sprintf("[✓] 成功解压 %s", filepath.Base(chunk)))
		}

		ea.addLog("")
	}

	// 验证 MD5
	if verifyMD5 {
		ea.addLog("[INFO] 开始验证 MD5...")
		ea.addLog("")

		verifySuccess, verified, failed := extractor.VerifyMD5(prefix)
		ea.addLog("")
		ea.addLog(fmt.Sprintf("[INFO] MD5 验证完成: %d 个通过, %d 个失败", verified, failed))

		if !verifySuccess && failed > 0 {
			success = false
		}
	}

	// 显示结果
	ea.addLog("============================================================")
	if success {
		ea.addLog("[✓] 解压完成！")
		ea.addLog(fmt.Sprintf("[✓] 所有文件已解压到: %s", outputDir))
		
		// 在主线程中更新UI
		ea.progressBar.SetValue(1.0)
		ea.progressLabel.SetText("✅ 完成")
		
		// 简洁的成功对话框
		successLabel := widget.NewLabel("解压完成！")
		successDialog := dialog.NewCustom("", "确定", 
			container.NewPadded(
				container.NewCenter(successLabel),
			), ea.window)
		successDialog.Resize(fyne.NewSize(250, 100))
		successDialog.Show()
	} else {
		ea.addLog("[✗] 解压过程中出现错误")
		
		// 在主线程中更新UI
		ea.progressLabel.SetText("出现错误")
		errorLabel := widget.NewLabel("解压失败，请查看日志")
		errorDialog := dialog.NewCustom("", "确定",
			container.NewPadded(
				container.NewCenter(errorLabel),
			), ea.window)
		errorDialog.Resize(fyne.NewSize(250, 100))
		errorDialog.Show()
	}
}

// Run 运行应用程序
func (ea *ExtractorApp) Run() {
	ea.SetupUI()
	prefixes, _ := AutoDetectPrefixes()
	if len(prefixes) > 0 {
		ea.addLog(fmt.Sprintf("[✓] 自动检测到前缀: %s", strings.Join(prefixes, ", ")))
	} else {
		ea.addLog("[!] 未检测到数据包文件（*_*.zip）")
	}
	ea.window.ShowAndRun()
}

func main() {
	app := NewExtractorApp()
	app.Run()
}

File: extract_tools/extract_tool_windows.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Windows数据包解压工具 v1.0
功能：
  - GUI界面，双击即可运行
  - 自动检测数据包前缀（查找*_001.zip）
  - 解压所有ZIP文件
  - 验证MD5完整性
  - Windows原生GUI支持

使用方法：
  1. 将本脚本放在与ZIP文件相同的目录
  2. 直接双击运行
  3. 选择输出目录和MD5验证选项
  4. 点击"开始解压"即可

编译为EXE：
  pyinstaller --onefile --windowed --icon=icon.ico extract_tool_windows.py
"""

import os
import sys
import json
import hashlib
import zipfile
from pathlib import Path
from typing import List, Tuple, Optional
import tkinter as tk
from tkinter import ttk, filedialog, messagebox, scrolledtext
import threading
import time


class DataExtractor:
    """数据解压器"""
    
    def __init__(self, prefix: str, output_dir: str):
        self.prefix = prefix
        self.output_dir = Path(output_dir)
        self.chunks = []
        self.log_messages = []
    
    def add_log(self, message: str):
        """记录日志"""
        self.log_messages.append(message)
    
    def find_chunks(self) -> List[str]:
        """查找所有ZIP分块文件"""
        current_dir = Path.cwd()
        chunks = []
        
        # 查找所有matching的zip文件
        for file in sorted(current_dir.glob(f"{self.prefix}_*.zip")):
            chunks.append(str(file))
        
        self.chunks = chunks
        return chunks
    
    @staticmethod
    def calculate_md5(file_path: Path) -> str:
        """计算文件MD5"""
        hash_md5 = hashlib.md5()
        with open(file_path, "rb") as f:
            for chunk in iter(lambda: f.read(4096), b""):
                hash_md5.update(chunk)
        return hash_md5.hexdigest()
    
    def extract_chunk(self, chunk_file: str, progress_callback=None) -> bool:
        """解压单个chunk文件"""
        try:
            chunk_path = Path(chunk_file)
            self.add_log(f"[INFO] 解压 {chunk_path.name}...")
            
            with zipfile.ZipFile(chunk_path, 'r') as zf:
                # 获取总文件数用于进度条
                file_list = zf.namelist()
                total_files = len(file_list)
                
                for idx, file_info in enumerate(file_list):
                    try:
                        zf.extract(file_info, self.output_dir)
                        if progress_callback:
                            progress_callback(idx + 1, total_files)
                    except Exception as e:
                        self.add_log(f"[WARN] 解压文件 {file_info} 时出错: {e}")
            
            self.add_log(f"[✓] 成功解压 {chunk_path.name}")
            return True
        except Exception as e:
            self.add_log(f"[✗] 解压失败 {chunk_file}: {e}")
            return False
    
    def extract_all(self, progress_callback=None) -> bool:
        """解压所有文件"""
        if not self.chunks:
            self.add_log("[✗] 未找到ZIP文件")
            return False
        
        failed = 0
        for idx, chunk in enumerate(self.chunks):
            if not self.extract_chunk(chunk, progress_callback):
                failed += 1
        
        return failed == 0
    
    def verify_md5(self, prefix: str, progress_callback=None) -> Tuple[bool, int, int]:
        """验证MD5完整性"""
        md5_file = Path(prefix + ".md5")
        
        if not md5_file.exists():
            self.add_log(f"[!] MD5文件不存在: {md5_file}")
            return True, 0, 0  # 不认为是错误
        
        verified_count = 0
        failed_count = 0
        
        self.add_log(f"[INFO] 开始MD5验证...")
        
        try:
            with open(md5_file, 'r', encoding='utf-8') as f:
                lines = [line.strip() for line in f if line.strip() and not line.startswith('#')]
            
            total_lines = len(lines)
            
            for idx, line in enumerate(lines):
                if progress_callback:
                    progress_callback(idx + 1, total_lines, "MD5验证")
                
                parts = line.split(None, 1)
                if len(parts) != 2:
                    continue
                
                expected_hash = parts[0]
                filename = parts[1]
                
                file_path = Path(filename)
                
                # 如果是相对路径，相对于输出目录
                if not file_path.is_absolute():
                    file_path = self.output_dir / filename
                
                if not file_path.exists():
                    self.add_log(f"[!] 文件未找到: {filename}")
                    failed_count += 1
                    continue
                
                # 计算实际MD5
                actual_hash = self.calculate_md5(file_path)
                
                if expected_hash.lower() == actual_hash.lower():
                    self.add_log(f"[✓] {filename}")
                    verified_count += 1
                else:
                    self.add_log(f"[✗] {filename} (MD5不匹配!)")
                    self.add_log(f"    期望: {expected_hash}")
                    self.add_log(f"    实际: {actual_hash}")
                    failed_count += 1
        
        except Exception as e:
            self.add_log(f"[✗] MD5验证失败: {e}")
            return False, verified_count, failed_count
        
        self.add_log(f"[INFO] MD5验证完成: {verified_count}个通过, {failed_count}个失败")
        
        return failed_count == 0, verified_count, failed_count


class ExtractorGUI:
    """解压工具GUI界面"""
    
    def __init__(self, root):
        self.root = root
        self.root.title("数据包解压工具 v1.0")
        self.root.geometry("600x700")
        self.root.resizable(False, False)
        
        # 设置中文字体
        try:
            self.font_normal = ("Microsoft YaHei", 10)
            self.font_title = ("Microsoft YaHei", 12, "bold")
            self.font_mono = ("Courier New", 9)
        except:
            self.font_normal = ("Arial", 10)
            self.font_title = ("Arial", 12, "bold")
            self.font_mono = ("Courier New", 9)
        
        self.extractor = None
        self.extraction_thread = None
        self.is_extracting = False
        
        self.setup_ui()
        self.auto_detect_prefix()
    
    def setup_ui(self):
        """设置UI界面"""
        # 标题
        title_frame = ttk.Frame(self.root)
        title_frame.pack(pady=10, padx=10, fill='x')
        
        title_label = tk.Label(
            title_frame,
            text="数据包解压工具",
            font=self.font_title,
            fg="#0066cc"
        )
        title_label.pack()
        
        # 参数设置框架
        param_frame = ttk.LabelFrame(self.root, text="解压参数", padding=10)
        param_frame.pack(pady=10, padx=10, fill='x')
        
        # 前缀选择
        ttk.Label(param_frame, text="数据包前缀:", font=self.font_normal).grid(row=0, column=0, sticky='w', pady=5)
        self.prefix_var = tk.StringVar()
        self.prefix_combo = ttk.Combobox(
            param_frame,
            textvariable=self.prefix_var,
            font=self.font_normal,
            width=40,
            state='readonly'
        )
        self.prefix_combo.grid(row=0, column=1, sticky='w', padx=5, pady=5)
        
        # 输出目录选择
        ttk.Label(param_frame, text="输出目录:", font=self.font_normal).grid(row=1, column=0, sticky='w', pady=5)
        output_frame = ttk.Frame(param_frame)
        output_frame.grid(row=1, column=1, sticky='ew', padx=5, pady=5)
        
        self.output_var = tk.StringVar(value=str(Path.cwd()))
        output_entry = ttk.Entry(
            output_frame,
            textvariable=self.output_var,
            font=self.font_normal,
            width=35
        )
        output_entry.pack(side='left', fill='x', expand=True)
        
        browse_btn = ttk.Button(
            output_frame,
            text="浏览",
            command=self.browse_output_dir,
            width=10
        )
        browse_btn.pack(side='left', padx=5)
        
        # 验证选项
        self.verify_var = tk.BooleanVar(value=False)
        verify_check = ttk.Checkbutton(
            param_frame,
            text="验证MD5完整性（可选）",
            variable=self.verify_var,
            font=self.font_normal
        )
        verify_check.grid(row=2, column=0, columnspan=2, sticky='w', pady=5)
        
        # 分割线
        ttk.Separator(self.root, orient='horizontal').pack(fill='x', pady=10)
        
        # 日志框架
        log_frame = ttk.LabelFrame(self.root, text="执行日志", padding=5)
        log_frame.pack(pady=10, padx=10, fill='both', expand=True)
        
        self.log_text = scrolledtext.ScrolledText(
            log_frame,
            font=self.font_mono,
            height=15,
            width=70,
            bg="#f5f5f5",
            fg="#333333"
        )
        self.log_text.pack(fill='both', expand=True)
        
        # 进度条
        self.progress_var = tk.DoubleVar()
        self.progress_bar = ttk.Progressbar(
            self.root,
            variable=self.progress_var,
            maximum=100,
            mode='determinate'
        )
        self.progress_bar.pack(pady=5, padx=10, fill='x')
        
        self.progress_label = tk.Label(
            self.root,
            text="就绪",
            font=self.font_normal,
            fg="#666666"
        )
        self.progress_label.pack()
        
        # 按钮框架
        btn_frame = ttk.Frame(self.root)
        btn_frame.pack(pady=10, padx=10, fill='x')
        
        self.extract_btn = ttk.Button(
            btn_frame,
            text="开始解压",
            command=self.start_extraction,
            width=15
        )
        self.extract_btn.pack(side='left', padx=5)
        
        clear_btn = ttk.Button(
            btn_frame,
            text="清空日志",
            command=self.clear_log,
            width=15
        )
        clear_btn.pack(side='left', padx=5)
        
        exit_btn = ttk.Button(
            btn_frame,
            text="退出",
            command=self.root.quit,
            width=15
        )
        exit_btn.pack(side='left', padx=5)
    
    def auto_detect_prefix(self):
        """自动检测数据包前缀"""
        current_dir = Path.cwd()
        prefixes = set()
        
        # 查找所有_001.zip文件
        for file in current_dir.glob("*_001.zip"):
            prefix = file.name[:-8]  # 去掉_001.zip
            prefixes.add(prefix)
        
        if prefixes:
            prefixes_list = sorted(list(prefixes))
            self.prefix_combo['values'] = prefixes_list
            self.prefix_combo.set(prefixes_list[0])
            self.add_log(f"[✓] 自动检测到前缀: {', '.join(prefixes_list)}")
        else:
            self.prefix_combo['values'] = []
            self.add_log("[!] 未检测到数据包文件（*_001.zip）")
    
    def browse_output_dir(self):
        """浏览输出目录"""
        dir_path = filedialog.askdirectory(
            title="选择输出目录",
            initialdir=self.output_var.get()
        )
        if dir_path:
            self.output_var.set(dir_path)
    
    def add_log(self, message: str):
        """添加日志消息"""
        self.log_text.insert('end', message + '\n')
        self.log_text.see('end')
        self.root.update_idletasks()
    
    def clear_log(self):
        """清空日志"""
        self.log_text.delete('1.0', 'end')
    
    def update_progress(self, current: int, total: int, stage: str = "解压"):
        """更新进度条"""
        if total > 0:
            percentage = (current / total) * 100
            self.progress_var.set(percentage)
            self.progress_label.config(text=f"{stage}: {current}/{total}")
        self.root.update_idletasks()
    
    def start_extraction(self):
        """开始解压"""
        prefix = self.prefix_var.get()
        output_dir = self.output_var.get()
        verify_md5 = self.verify_var.get()
        
        if not prefix:
            messagebox.showerror("错误", "请选择数据包前缀")
            return
        
        if not output_dir:
            messagebox.showerror("错误", "请选择输出目录")
            return
        
        # 禁用按钮
        self.is_extracting = True
        self.extract_btn.config(state='disabled')
        self.prefix_combo.config(state='disabled')
        
        # 在线程中执行解压
        self.extraction_thread = threading.Thread(
            target=self._extraction_worker,
            args=(prefix, output_dir, verify_md5),
            daemon=True
        )
        self.extraction_thread.start()
    
    def _extraction_worker(self, prefix: str, output_dir: str, verify_md5: bool):
        """解压工作线程"""
        try:
            # 创建输出目录
            output_path = Path(output_dir)
            output_path.mkdir(parents=True, exist_ok=True)
            
            self.add_log(f"[INFO] 开始解压...")
            self.add_log(f"[INFO] 数据包前缀: {prefix}")
            self.add_log(f"[INFO] 输出目录: {output_dir}")
            self.add_log("")
            
            # 创建解压器
            self.extractor = DataExtractor(prefix, output_dir)
            
            # 查找文件
            chunks = self.extractor.find_chunks()
            self.add_log(f"[INFO] 找到 {len(chunks)} 个分块文件:")
            for chunk in chunks:
                self.add_log(f"  - {Path(chunk).name}")
            self.add_log("")
            
            if not chunks:
                self.add_log("[✗] 没有找到ZIP文件")
                messagebox.showerror("错误", "没有找到数据包文件")
                return
            
            # 解压所有文件
            success = True
            for idx, chunk in enumerate(chunks):
                self.add_log(f"[INFO] 正在解压第 {idx + 1}/{len(chunks)} 个文件...")
                
                def progress_callback(current, total):
                    self.update_progress(current, total, f"解压文件 ({idx + 1}/{len(chunks)})")
                
                if not self.extractor.extract_chunk(chunk, progress_callback):
                    success = False
                
                self.add_log("")
            
            # 验证MD5
            if verify_md5:
                self.add_log("[INFO] 开始验证MD5...")
                self.add_log("")
                
                def md5_progress(current, total, stage=""):
                    self.update_progress(current, total, stage)
                
                verify_success, verified, failed = self.extractor.verify_md5(prefix, md5_progress)
                self.add_log("")
                
                if not verify_success and failed > 0:
                    success = False
            
            # 显示结果
            self.add_log("=" * 60)
            if success:
                self.add_log("[✓] 解压完成！")
                self.add_log(f"[✓] 所有文件已解压到: {output_dir}")
                self.progress_var.set(100)
                self.progress_label.config(text="完成")
                messagebox.showinfo("成功", "解压完成！")
            else:
                self.add_log("[✗] 解压过程中出现错误")
                self.progress_label.config(text="出现错误")
                messagebox.showerror("失败", "解压过程中出现错误，请查看日志")
        
        except Exception as e:
            self.add_log(f"[✗] 异常错误: {e}")
            messagebox.showerror("异常", f"发生异常: {e}")
        
        finally:
            # 恢复按钮
            self.is_extracting = False
            self.extract_btn.config(state='normal')
            self.prefix_combo.config(state='readonly')


def main():
    """主函数"""
    root = tk.Tk()
    
    # 设置图标（如果存在）
    try:
        # 尝试加载图标
        icon_path = Path(__file__).parent / "icon.ico"
        if icon_path.exists():
            root.iconbitmap(str(icon_path))
    except:
        pass
    
    gui = ExtractorGUI(root)
    root.mainloop()


if __name__ == "__main__":
    main()
